### DISCUSSION TITLE
VDO Latency

### DISCUSSION THREAD

**Discussion by Cacsjep** [AxisCommunications/discussions#157] [ACAP]
**Hey Axis Team,**

We are currently developing a viewing client for industrial applications and have reached a point where VDO latency is a major value in the chain. After several tests we came to the conclusion that it seems that VDO itself has a high latency when we assume a near real-time display. 

Our rendering plus decompression for JPEG takes less than 25ms. With the VDO latency, this results in an end to end latency of about 50ms. We have tested and validated this in several different tests. 

However, this has been an experiment to see how far you can reduce the end to end latency. Our final solution is a browser based one. To have a better cross platform solution. Here we have an end to end latency of about 65ms.

If we use VDO_TIMESTAMP_MONO_CLIENT_CAPTURE_DIFF and compare the timestamp, the image already has a latency around 22ms. So before we do anything with the image we already have this latency. That would be around 33% of the total end to end latency with our browser solution.

We also see that there is a tool on the camera (vdo-client) that shows similar values for latency. I also checked this latency by placing the camera in front of a 240Hz monitor with a millisecond timer and then comparing the image timestamp with the timestamp on the monitor, in the captured frame. (Saved directly on camera with <ts>.jpeg)

Setting buffer.count on the VDO Map does not make any diffrence, 

Testing vdo with the existing axis tool. (AXIS Q1615 Mk III 100FPS, ARTPEC-7)
`/usr/bin/vdo-client -n 20 -t yuv -o /dev/null -s -c <compression> | grep vdo.latency.`
![Artpec-7](https://user-images.githubusercontent.com/98898150/212040863-64efeca3-1e3d-4be9-b8e6-009e966aa859.PNG)

Testing vdo with the existing axis tool. (AXIS Q3536 Mk 60FPS, ARTPEC-8)
![Artpec-8](https://user-images.githubusercontent.com/98898150/212048617-29da6a79-f1ae-486f-ae66-9737c0fdd420.PNG)

I did our tests with Low Latency mode activated and all image optimizations turned off. Also no rotation, overlay, Privacy set. etc..

The only really thing to reduce VDO latency is when I decrease the resolution, but thats not an option for us.
Also tests show that with higher resolutions, VDO latency is increasing, but not for YUV, so it looks to me that maybe compression is here the factor of latency.

Of course I can imagine that there are a lot of things going on in the background, but maybe you have an idea or info for me how I can reduce VDO latency without changing the resolution even 10ms savings would be great, or maybe tune some internal compression params for jpeg. If this is not possible, then it is just so, then this is so to speak state of art.

Thanks
Br Christoph

**Comment by Cacsjep** [comment]
Support has answering me that question =)

**Comment by Vrzi** [comment]
Hi, Can you share the answer from support?

**Reply by Cacsjep** [reply]
Axis Support - 13 jan 2023 15:21
HI Christoph,

thank you for your patience.

Developer's feedback to your query:

Total HW latency: ~23ms

— Breakdown —

Bayer->minimal image processing->yuv/rgb: ~11ms
Rest of HW pipeline (scale/enocde etc): ~12ms
The vdo framework itself adds almost no latency.

vdo-client -t jpeg -u client_server_diff
vdo framework latency: ~150-300µs
We tried to use Power line frequency 60 -> 120fps instead of 100fps
120fps jpeg latency
latency sensor->vdo: ~67ms
120fps jpeg + low-latency mode
latency sensor->vdo: ~23ms
lower-resolution 960x540 jpeg: ~17ms

Depending on your intended usecase... Rather than using a low quality 50% jpeg at 1080p you could possibly use a high quality jpeg at 960x540 and upscale at the PC side to 1080p?

There is no measurable difference between 11.x and 10.x AxOS.
vdo-client -t jpeg 960 5

Best regards,
Axis TIP Helpdesk
